### PR TITLE
Burning for external contracts

### PR DESCRIPTION
**Pull request by lbelyaev**
IBurnRedeemable interface
Burn confirmation hook

### CODE REVIEW

**Review by nibty — APPROVED** [review]

File: contracts/XENCrypto.sol [hunk 1]
```diff
@@ -403,4 +409,20 @@ contract XENCrypto is Context, IRankedMintingToken, IStakingToken, ERC20("XEN Cr
...
+        require(amount > XEN_MIN_BURN, "Burn: Below min limit");
+        require(
+            IERC165(_msgSender()).supportsInterface(type(IBurnRedeemable).interfaceId),
+            "Burn: not a supported contract"
+        );
+
+        _spendAllowance(user, _msgSender(), amount);
+        _burn(user, amount);
+        userBurns[user] += amount;
+        IBurnRedeemable(_msgSender()).onTokenBurned(user, amount);
```
Comment (mastergaurang94) [inline contracts/XENCrypto.sol:426]: This is nice! I didn't realize hooks could be made like this..

File: contracts/XENCrypto.sol [hunk 2]
```diff
@@ -403,4 +409,20 @@ contract XENCrypto is Context, IRankedMintingToken, IStakingToken, ERC20("XEN Cr
...
+     */
+    function burn(address user, uint256 amount) public {
+        require(amount > XEN_MIN_BURN, "Burn: Below min limit");
+        require(
+            IERC165(_msgSender()).supportsInterface(type(IBurnRedeemable).interfaceId),
+            "Burn: not a supported contract"
+        );
+
+        _spendAllowance(user, _msgSender(), amount);
+        _burn(user, amount);
```
Comment (mastergaurang94) [inline contracts/XENCrypto.sol:424]: Curious: How do external contracts get approval? By asking Xen?

File: test/XENCrypto-Burn.test.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+    it('Should not allow reentrancy calls; should maintain original state', async () => {
+        await assert.doesNotReject(() => token.approve(badBurner.address, balance / 2n, {from: accounts[1]}));
+        await truffleAssert.fails(
+            badBurner.exchangeTokens(balance / 100n, {from: accounts[1]})
+        )
+        const xenBalance = await token.balanceOf(accounts[1], {from: accounts[1]}).then(toBigInt)
+        const otherBalance = await burner.balanceOf(accounts[1], {from: accounts[1]}).then(toBigInt)
+        const badBalance = await badBurner.balanceOf(accounts[1], {from: accounts[1]}).then(toBigInt)
+        // console.log('bad', badBalance)
```
Comment (mastergaurang94) [inline test/XENCrypto-Burn.test.js:92]: console log

**Review by mastergaurang94 — APPROVED** [review]
+1

**Pull request merged**